nvflare/app_common/ccwf/client_ctl.py

```diff
@@ -0,0 +1,630 @@
+
+        # process request from prev client
+        self.log_info(fl_ctx, f"Got Learn request from {sender}")
+
+        if self.learn_task and not self.allow_busy_task:
+            # should never happen!
+            self.log_error(fl_ctx, f"got Learn request from {sender} while I'm still busy!")
+            self.update_status(action="process_learn_request", error=ReturnCode.EXECUTION_EXCEPTION)
+            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+
+        self.log_info(fl_ctx, f"accepted learn request from {sender}")
+        self.set_learn_task(task_data=request, fl_ctx=fl_ctx)
+        return make_reply(ReturnCode.OK)
+
+    def send_learn_task(self, targets: list, request: Shareable, fl_ctx: FLContext) -> bool:
+        self.log_info(fl_ctx, f"sending learn task to clients {targets}")
+        request.set_header(AppConstants.NUM_ROUNDS, self.get_config_prop(AppConstants.NUM_ROUNDS))
+
+        task = Task(
+            name=self.do_learn_task_name,
+            data=request,
+            timeout=int(self.learn_task_ack_timeout),
+        )
+
+        resp = self.broadcast_and_wait(
+            task=task,
+            targets=targets,
+            min_responses=len(targets),
+            fl_ctx=fl_ctx,
+        )
+
+        assert isinstance(resp, dict)
+        for t in targets:
+            reply = resp.get(t)
+            if not isinstance(reply, Shareable):
+                self.log_error(fl_ctx, f"failed to send learn request to client {t}")
+                self.log_error(fl_ctx, f"reply must be Shareable but got {type(reply)}")
+                self.update_status(action="send_learn_task", error=ReturnCode.EXECUTION_EXCEPTION)
+                return False
+
+            rc = reply.get_return_code(ReturnCode.OK)
+            if rc != ReturnCode.OK:
+                self.log_error(fl_ctx, f"bad response for learn request from client {t}: {rc}")
+                self.update_status(action="send_learn_task", error=rc)
+                return False
+        return True
+
+    def execute_train(self, data: Shareable, fl_ctx: FLContext, abort_signal: Signal) -> Shareable:
+        current_round = data.get_header(AppConstants.CURRENT_ROUND)
+
+        self.log_info(fl_ctx, f"started training round {current_round}")
+        try:
+            result = self.learn_executor.execute(self.learn_task_name, data, fl_ctx, abort_signal)
+        except:
+            self.log_exception(fl_ctx, f"trainer exception: {secure_format_traceback()}")
+            result = make_reply(ReturnCode.EXECUTION_EXCEPTION)
+        self.log_info(fl_ctx, f"finished training round {current_round}")
+
+        # make sure to include cookies in result
+        cookie_jar = data.get_cookie_jar()
+        result.set_cookie_jar(cookie_jar)
+        result.set_header(AppConstants.CURRENT_ROUND, current_round)
+        result.add_cookie(AppConstants.CONTRIBUTION_ROUND, current_round)  # to make model selector happy
+        return result
```

Comment:
Should it call "execute_learn_task"?
And should this be moved inside the "do_learn_task" method in child classes?

Reply:
Good suggestion for renaming to execute_learn_task.
Not sure what you mean by moving it inside of do_learn_task. But each subclass has different impl of do_learn_task (before and after calling execute_learn_task".